src/layer_env.rs

```diff
@@ -0,0 +1,528 @@
+                entry.name == key.as_ref()
+                    && entry.modification_behavior == ModificationBehavior::Delimiter
+            })
+            .map(|entry| entry.value.clone())
+            .unwrap_or(OsString::new())
+    }
+
+    fn read_from_env_dir(path: impl AsRef<Path>) -> Result<Self, std::io::Error> {
+        let mut layer_env = Self::empty();
+
+        for dir_entry in fs::read_dir(path.as_ref())? {
+            let path = dir_entry?.path();
+
+            // Rely on the Rust standard library for splitting stem and extension. Since paths
+            // are not necessarily UTF-8 encoded, this is not as trivial as it might look like.
+            // Think twice before changing this.
+            let file_name_stem = path.file_stem();
+            let file_name_extension = path.extension();
+
+            // The CNB spec explicitly states:
+            //
+            // > File contents MUST NOT be evaluated by a shell or otherwise modified before
+            // > inclusion in environment variable values.
+            // > https://github.com/buildpacks/spec/blob/a9f64de9c78022aa7a5091077a765f932d7afe42/buildpack.md#provided-by-the-buildpacks
+            //
+            // This should include parsing the contents with an assumed charset and later emitting
+            // the raw bytes of that encoding as it might change the actual data. Since this is not
+            // explicitly written in the spec, we read through the the reference implementation and
+            // determined that it also treats the file contents as raw bytes.
+            // See: https://github.com/buildpacks/lifecycle/blob/a7428a55c2a14d8a37e84285b95dc63192e3264e/env/env.go#L73-L106
+            use std::os::unix::ffi::OsStringExt;
+            let file_contents = OsString::from_vec(fs::read(&path)?);
+
+            if let Some(file_name_stem) = file_name_stem {
+                let r#type = match file_name_extension {
+                    None => {
+                        // TODO: This is different for CNB API versions > 0.5:
+                        // https://github.com/buildpacks/lifecycle/blob/a7428a55c2a14d8a37e84285b95dc63192e3264e/env/env.go#L66-L71
+                        Some(ModificationBehavior::Override)
+                    }
+                    Some(file_name_extension) => match file_name_extension.to_str() {
+                        Some("append") => Some(ModificationBehavior::Append),
+                        Some("default") => Some(ModificationBehavior::Default),
+                        Some("delim") => Some(ModificationBehavior::Delimiter),
+                        Some("override") => Some(ModificationBehavior::Override),
+                        Some("prepend") => Some(ModificationBehavior::Prepend),
+                        // Note: This IS NOT the case where we have no extension. This handles
+                        // the case of an unknown or non-UTF-8 extension.
+                        Some(_) | None => None,
```

Comment:
When do we think this "unknown extension" scenario can occur? libcnb.rs will only be parsing the env dirs created by the current buildpack invocation, right? ie: Both earlier buildpacks in the build (whose values get set in in the system env by lifecycle) and cached previous builds won't make any difference? Which would mean this function is only ever having to parse values it wrote itself, unless someone is manually editing the env dir before calling this libcnb function?

If so, then does returning an Error make more sense here? (Given this isn't going to affect backwards compatibility or anything)

Reply:
^

Reply:
Like I said in [another comment](https://github.com/Malax/libcnb.rs/pull/30#discussion_r699079356), I don't think we should make any assumptions about the file system contents for various reasons. Let me know if you want me to talk about this specific instance more or if the link to the other comment is enough. :)